presto-main/src/main/java/com/facebook/presto/operator/SpatialJoinOperator.java

```diff
@@ -0,0 +1,244 @@
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return outputTypes;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return pagesSpatialIndexFuture.isDone() && !pageBuilder.isFull() && probe == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        verify(probe == null);
+        probe = page;
+        probePosition = 0;
+
+        joinAddresses = null;
+        joinAddressIndex = 0;
+    }
+
+    private void processProbe()
+    {
+        verify(probe != null);
+
+        PagesSpatialIndex pagesSpatialIndex = getDone(pagesSpatialIndexFuture);
+        for (int position = probePosition; position < probe.getPositionCount(); position++) {
+            if (this.joinAddresses == null) {
+                joinAddresses = pagesSpatialIndex.findJoinAddresses(position, probe, probeGeometryChannel, spatialRelationshipTest);
+                joinAddressIndex = -1;
+            }
+
+            for (int joinAddressIndex = this.joinAddressIndex + 1; joinAddressIndex < joinAddresses.size(); joinAddressIndex++) {
+                if (pageBuilder.isFull()) {
+                    probePosition = position;
+                    this.joinAddressIndex = joinAddressIndex;
+                    return;
+                }
+
+                long joinAddress = joinAddresses.get(joinAddressIndex);
+
+                pageBuilder.declarePosition();
+                int outputChannelOffset = 0;
+                for (int outputIndex : probeOutputChannels) {
+                    Type type = probeTypes.get(outputIndex);
+                    Block block = probe.getBlock(outputIndex);
+                    type.appendTo(block, position, pageBuilder.getBlockBuilder(outputChannelOffset));
+                    outputChannelOffset++;
+                }
+                pagesSpatialIndex.appendTo(joinAddress, pageBuilder, outputChannelOffset);
+            }
+
+            this.joinAddresses = null;
+        }
+
+        this.probe = null;
+        this.probePosition = 0;
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (!pageBuilder.isFull() && probe != null) {
+            processProbe();
+        }
+
+        if (pageBuilder.isFull()) {
+            Page page = pageBuilder.build();
+            pageBuilder.reset();
+            return page;
+        }
+
+        if (finishing) {
+            Page page = null;
+            if (!pageBuilder.isEmpty()) {
+                page = pageBuilder.build();
+                pageBuilder.reset();
+            }
+            finished = true;
+            return page;
+        }
+
+        return null;
+    }
+
+    @Override
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finished;
+    }
```

Comment:
When will the `pagesSpatialIndex` kept in `pagesSpatialIndexFuture` be freed?

Reply:
I assumed these will be freed automatically once the `SpatialJoinOperator` is finished. If this is not the case, would you help me figure out what I need to do?